src/render/tutorials/api-gateway-security/2_0.html.md

```diff
@@ -0,0 +1,949 @@
+      throw new IllegalArgumentException("Configuration for Google Auth must include code, clientId, clientSecret and redirectUri");
+    }
+
+    return webClient.postAbs(TOKEN_URL)
+        .setQueryParam("code", code)
+        .setQueryParam("client_id", clientId)
+        .setQueryParam("client_secret", clientSecret)
+        .setQueryParam("redirect_uri", redirectUri)
+        .setQueryParam("grant_type", "authorization_code")
+        .putHeader("Content-Length", "0");
+  }
+
+  private Single<HttpResponse<Buffer>> fetchUserInfo(String accessToken) {
+    return webClient.getAbs(USERINFO_URL)
+        .bearerTokenAuthentication(accessToken)
+        .rxSend();
+  }
+
+}
+
+```
+
+Let's take a look at this file step by step:
+
+1. The constructor is quite straightforward, we create a new [`webClient`](https://vertx.io/docs/vertx-web-client/java/) and save the config for later use.
+2. When a request hits our handler, the `handle` method is invoked. We try to obtain the authorization code from the request.
+3. Then we try to `exchangeCodeForToken`. The first step is to `prepareTokenRequest`. This time we have to provide the `code`, `client_id`, `client_secret`, `redirect_uri`, `grant_type` parameters and `Content-Length` header (this will be a `POST` request). For more details on those parameters see [this documentation](https://developers.google.com/identity/protocols/OAuth2WebServer#exchange-authorization-code). We send the request using `rxSend` method (for more info about Vert.x's RxJava 2 API check [this documentation](https://vertx.io/docs/vertx-web-client/java/#_rxjava_2_api)) and map the response to the acquired access token.
+4. We map the resulting observable with the token to `fetchUserInfo` whick creates a `GET` request to Google API with bearer token authentication and sends it.
+5. Finally we subscribe for the whole processing chain and end our response with a personalized hello message!
+
+As always, let's not forget to register our factories in the `resources/META_INF/services` directory!
+`io.knotx.server.api.handler.RoutingHandlerFactory`
+```
+io.knotx.examples.security.routing.GoogleLoginRedirectHandlerFactory
+io.knotx.examples.security.routing.OAuth2HelloHandlerFactory
+```
+
+### Build & Run
+
+Now let's see how it works.
+
+First, let's build the project (as mentioned before, remember to stop all the running containers!):
+
+```
+$ gradlew clean build-docker
+```
+
+and run it:
+
+```
+$ docker run -p8092:8092 knotx-example/secure-api-gateway
+```
+
+Now let's type in `http://localhost:8092/api/secure/login` in your browser.
+
+![Select Google Account](/img/blog/api-gateway-security/security-google-select-account.png)
+
+Select your account.
+
+![Select Google Account](/img/blog/api-gateway-security/security-google-hello-message.png)
+
+And there is our personalized hello message!
+
+---
+
+You can find full project implementation [here](https://github.com/Knotx/knotx-example-project/tree/master/api-gateway/security).
```

Comment:
How about some summary of what we actually did and what we learned and why !x is awesome etc

Reply:
I've provided some, let me know if it is satisfactory